Mediatek Audio DSPs: 8186/8188/8196

andyross opened the pull request:
This is getting hot again so re-pushing a stale PR with a few bits added.  Support for Zephyr builds for the Audio DSPs on MT8186, MT8188 and MT8196 SOCs.

andyross committed:
soc: boards: Add Mediatek MT8186 and MT8188 audio DSPs

These are very similar devices to mt8195, minimal changes needed
beyond boilerplate configuration.

In the process, this reworks the board/soc layout to be HWMv2
compliant, with "adsp" becoming a CPU cluster beneath the SOC.  So the
name of the boards to west become e.g. "mt8195/mt8195/adsp" (which can
be shortened to "mt8195//adsp" if desired).

Note that the cpuclk driver is not yet ported, it works only with 8195
(the clocking/power architecture seems similar between the parts, but
the graph of wells and clocks is different and historically these have
been three separate drivers in SOF).  The biggest changes are in the
image/loader scripts, which needed some rework for cross-device
portability.

Signed-off-by: Andy Ross <andyross@google.com>

andyross committed:
arch: Let CONFIG_IRQ_OFFLOAD_NESTED be overridable

This is currently selected by the arch, but some devices (in paticular
xtensa ones which have configurable interrupt setups) might not
actually be able to effect an irq_offload() from within an ISR even
though we have code for it.

Make this a default and not a select so that lower layers can override
the setting.

Signed-off-by: Andy Ross <andyross@google.com>

andyross committed:
soc/mtk_adsp: Add default console hook

Wire the default printk output to the console at boot, just to be sure
we have stdio output good enough to get tests to pass.

Signed-off-by: Andy Ross <andyross@google.com>

andyross committed:
soc/mtk_adsp: Remove "msg" API

This is a feature of the 8195 DSP only, which is used only vestigially
by SOF to store data that nothing reads.  The Linux kernel on the
other side uses a shared driver for all 81xx devices, which does not
expose the feature.  It seems to work, but it's not worth maintaining
a driver in tree for legacy hardware that will never use it.

Signed-off-by: Andy Ross <andyross@google.com>

andyross committed:
boards/mtk_adsp: Add a doc page

Simple docs for this board family.  Not a lot of complexity currently.

Signed-off-by: Andy Ross <andyross@google.com>

andyross committed:
boards/mediatek: Add mt8196_adsp

Add Zephyr support for the Audio DSP on the MT8196 SOC.  This is a
very similar device to previous designs.  Most of this patch is just
DTS.

The biggest delta is the more complicated second level interrupt
controller, though it is still able to be represented using some
vaguely clever DTS config over the older intc_mtk_adsp driver.

Also the memory layout is slightly different, requiring a little
indirection to set the initial boot stack address and log output
buffer.  And the timer "irq_ack" register bits moved.

Signed-off-by: Andy Ross <andyross@google.com>

andyross committed:
soc/mtk_adsp: Add missing z_prep_c() prototype

The early boot function got renamed to a pseudo-standard "z_prep_c",
but this isn't an actual API and doesn't have a prototype in the
headers anywhere, so the compiler started whining about an undeclared
function.

Signed-off-by: Andy Ross <andyross@google.com>

andyross committed:
soc/mtk_adsp: Update gen_img.py address space rules

New platform has different mappings.  Auto-detect rather than parse
dts or similar, as this is is really just a simple format for testing.

Signed-off-by: Andy Ross <andyross@google.com>

andyross commented:
Add shotgun gaggle of Xtensa reviewers.

nordicjm commented on soc/mediatek/mtk_adsp/soc.yml:
```diff
@@ -1,6 +1,10 @@
           - name: mt8195_adsp
+      - name: mt818x_adsp
+        socs:
+          - name: mt8186_adsp
+          - name: mt8188_adsp
```
is the actual soc `mt8186` or is it `mt8186_adsp`? If the soc is `mt8186` then adsp should be a cpu cluster, not part of the soc

andyross replied:
These are bigcore chips with lots of subdevices intended for consumer Linux applications.  They'll never be able to share a DTS tree with the rest of the device in any practical sense.  Where we've dealt with similar DSPs in the past (c.f. soc/Intel/inetl_adsp) there's never been an attempt to shoehorn the extra hardware in and the "SOC" has always just represented the audio hardware.

I guess I'd prefer to keep that convention for simplicity and commonality, but if you really want the change can you clarify exactly how you think it should work?  I'm grepping around and not finding an obvious example of a CPU cluster to emulate.

andyross replied:
Ah... Yeah, I'm going to push back pretty hard on that.  Not only is it extra typing, it actually breaks the real feature that SOC_SERIES provides.   The 8188 and 8186 DSPs are actually the same Xtensa core and share almost all the configuration bits, but they live in rather different chips with different MMIO addresses.  Forcing them to be CPU clusters breaks this and I'd need to double the various #ifdef checks, etc...

And again, it's not like doing this "right" per the Zephyr hwmv2 specification is going to matter to anyone else using the chips, because they're big Linux application processors working in their own world with their own tooling and will never care about what we're doing here in Zephyr.  Really this is very much a hobgoblin-of-little-minds situation, IMHO.  I get that there's a space for complicated modeling of hardware environments, but we absolutely shouldn't be forcing it onto all users.

nordicjm replied:
the 8188 and 8186 would be separate SOCs, the part that would become a CPU cluster is the adsp, and no on the contrary it would be the same amount typing, if you have it as you do now, you have to do `west build -b <board>/mt8186_adsp` then this would instead be `west build -b <board>/mt8186/adsp` or if the board just has a single SOC and single CPU cluster it would actually be `west build -b <board>` without the SOC/CPU cluster on the end

tejlmand replied:
> or if the board just has a single SOC and single CPU cluster it would actually be west build -b <board> without the SOC/CPU cluster on the end.

actually the CPU cluster always needs to be defined, even if there is just a single.
Perhaps we should allow users to omit the CPU cluster if there is just a single cluster, but at the time we did HWMv2, then all SoC defining CPU clusters were all having at least two.

So the correct syntax would be: `west build -b <board>/mt8186/adsp` or `west build -b <board>//adsp`.

But based on discussion in this thread, allowing users to omit the CPU clusters when there is only a single makes sense.

tejlmand replied:
from what I can see in this PR then i'm aligned with @nordicjm on this one.

Having the `adsp` modeled as a CPU cluster actually makes this clearer that the adsp is part of something bigger, and where we are not supporting this bigger, because we are not modeling the other CPU clusters.

@nordicjm is not asking you to describe or implement the extra Zephyr unsupported CPU cluster.
And as it is also pointed out, then for the end-user, all which changes in having to use a `/` instead of `_` in `<board>/mt8186/adsp`.

 > Not only is it extra typing, it actually breaks the real feature that SOC_SERIES provides. The 8188 and 8186 DSPs are actually the same Xtensa core and share almost all the configuration bits, but they live in rather different chips with different MMIO addresses. Forcing them to be CPU clusters breaks this and I'd need to double the various #ifdef checks, etc...

I'm curious here ?
Can you explain exactly what extra typing will be needed and for who ?
(Only place I see is the req to always specify CPU cluster, as mentioned in https://github.com/zephyrproject-rtos/zephyr/pull/81154/commits/ef0e92784ff401778fcd303ff0c0fdb59f58e577#r1839968759, but I'd be hapy to address that.
As well as why you cannot share the common configuration bits ?

andyross replied:
OK, I give up.  Not worth a fight over what ultimately is just extra indirection in stuff that doesn't impact the build.

But to be clear: MT8186 and MT8188 are not really "family-related" devices.  I mean, no more so than they are to the other members of the series.  Different AP core counts, different market targeting, etc...   But they **happen** to share exactly the same Audio DSP design (modulo, again, stuff like bus addresses).  So I can very reasonably pretend that they share the same SOC series when looking only at audio.  That gets pretty strained when the config file starts to pretend it represents the whole chip (which it doesn't of course, and never will).

Now, sure, I can come up with my own abstraction for this relationship that isn't the one baked into hwmv2.  It's just, as I said, more typing.  For IMHO essentially zero value.

nashif replied:
interesting discussion and confusing as well.  I thought the cpucluster would be an ARM things and we would not have to deal with it anywhere.  Having read the documentation for this multiple time does not help much https://docs.zephyrproject.org/latest/glossary.html#term-CPU-cluster in the context of adsp:

![image](https://github.com/user-attachments/assets/8f19205b-9c8f-4265-be70-df9dfb9cc3bb)

andyross replied:
Yeah.  In this case the distinction only makes sense when viewing the whole chip.  So you might have (to choose a less obscure device) an "intel_arrow_lake" SOC, with a cluster of "lion_cove" application processors, a cluster of "intel_ace15" audio DSPs, and in theory stuff like the camera processor and PMC and CSME and anything else that could run Zephyr.

Except that in practice all anyone cares about with this submission is the single Audio DSP core,  as the AP's on these chips are running Linux and live in an entirely different tooling world.  Which is basically the way "ancient Zephyr" configured its devices, with no need to track board-vs-soc-vs-... But alas the "simple" design is no longer simple.

nashif replied:
yes, exactly. And bringing in SMP into the terminology does not help much. Based on the above someone will now ask us to change all the intel IP definitions we have as part of the SoC product as cpu clusters, the shoehorning the cpu cluster terminology on such usages is an overkill IMO.

nordicjm replied:
I mean the point of hwmv2 was to have all these properties defined properly, it's not only an ARM thing. If there is only one core in a device then it doesn't make sense to add the CPU cluster but if there's more than one core then it does, you originally added the original hwmv1 board with `_adsp` when you didn't have to. And as said, if you want a common Kconfig (which can be named anything really so long as it's not using a reserved hwmv2 name like `SOC_SERIES_*`) which 2 or more SoCs select for usage in Kconfig files or drivers to allow sharing of common drivers then you can do that. They don't need to be part of the same soc series even, it's just Kconfig after all

andyross replied:
FWIW, I think this gets to a conflicting definition for "proper".  You want all the boxes checked to fit the config into a scheme that's designed to match every possible device.  I just want to build audio firmware for a handful of devices, and know for a fact that no one is ever going to care about Zephyr configs for the big Cortex-A/X cores that happen to share space on the silicon layout.  Making me jump through hoops and understand a complicated scheme seems... "improper".

It's the Alan Kay quote, basically.  The ideal for an API design is "Simple things should be simple, complex things should be possible.".  hwmv2 makes everything complex, simple things stop being simple.

teburd replied:
Is it really just a cpu inside an SoC? In many cases these things are SoCs in their own right that just *happened* to be part of a larger part, with their own versioning, isa, cpu cluster, peripheral set, memory, etc.

nordicjm replied:
hwmv2 had input from and was approved by many people, including intel. I actually find it quite jarring that as the result of having ONE underscore turn to a slash in your build commands generates this much resistance, like that is literally it, which makes it fit into the hardware scheme that everyone in zephyr is using. If no-one adds support for them, no-one adds support for them, to say that no-one ever will is a bit of a crass overstatement given that there are "big" cortex-a cores supported in zephyr e.g. raspberry pi 5, seemingly people do care about running zephyr on them.

nordicjm commented on soc/mediatek/mtk_adsp/Kconfig.soc:
```diff
@@ -1,8 +1,34 @@
 # SPDX-License-Identifier: Apache-2.0
 
+config SOC_FAMILY_MTK_ADSP
+	bool "Mediatek MT8xxx Series Audio DSPs"
```
```suggestion
	bool
```

nordicjm commented on soc/mediatek/mtk_adsp/Kconfig.soc:
```diff
@@ -1,8 +1,34 @@
+	bool "Mediatek MT8xxx Series Audio DSPs"
+
+config SOC_SERIES_MT8195_ADSP
+	bool "Mediatek 8195 Audio DSP"
```
```suggestion
	bool
```

nordicjm commented on soc/mediatek/mtk_adsp/Kconfig.soc:
```diff
@@ -1,8 +1,34 @@
+	  Mediatek MT8195 Audio DSP
+
+config SOC_SERIES_MT818X_ADSP
+	bool "Mediatek 818x Audio DSPs"
```
```suggestion
	bool
```

nordicjm commented on boards/mediatek/mt8195_adsp/Kconfig.mt8195_adsp:
```diff
@@ -2,6 +2,5 @@
 # SPDX-License-Identifier: Apache-2.0
 
 config BOARD_MT8195_ADSP
-	select SOC_MT8195_ADSP
 	help
 	  Board with Mediatek MT8195 Audio DSP
```
??? this cannot work

andyross replied:
Clarify?  I mean, it builds OK for me.  The 8195 code is in-tree already and predates hwmv2.  My guess is this was left in during the transition and I'm just removing it for hygiene?

nordicjm commented on boards/mediatek/mt8188_adsp/mt8188_adsp.dts:
```diff
@@ -0,0 +1,82 @@
+/dts-v1/;
+/ {
+
+#address-cells = <1>;
```
missing indent?

nordicjm commented on boards/mediatek/mt8188_adsp/Kconfig.mt8188_adsp:
```diff
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_MT8188_ADSP
+	bool "Mediatek MT8188 Audio DSP"
```
```suggestion
```

nordicjm commented on boards/mediatek/mt8186_adsp/mt8186_adsp.dts:
```diff
@@ -0,0 +1,82 @@
+/dts-v1/;
+/ {
+
+#address-cells = <1>;
```
missing indent?

andyross replied:
Intentionally, as the only content in the file is in the outer ("/") scope, so it would have to indent the entire file.  Same reason you don't indent for namespace declarations in C++, basically.

nordicjm commented on boards/mediatek/mt8196_adsp/Kconfig.mt8196_adsp:
```diff
@@ -0,0 +1,6 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_MT8196_ADSP
+	bool "Mediatek MT8196 Audio DSP"
```
```suggestion
```

nordicjm commented on boards/mediatek/mt8196_adsp/mt8196_adsp.dts:
```diff
@@ -0,0 +1,115 @@
+/dts-v1/;
+/ {
+
+#address-cells = <1>;
```
missing indent?

nordicjm commented on soc/mediatek/mtk_adsp/Kconfig.soc:
```diff
@@ -16,6 +16,12 @@ config SOC_SERIES_MT818X_ADSP
 	  Mediatek MT818x Audio DSPs
 
+config SOC_SERIES_MT8196_ADSP
+	bool "Mediatek 8196 Audio DSP"
```
```suggestion
	bool
```

nordicjm reviewed (dismissed):
See query about cpu cluster first, cpu clusters should be used if you are targetting a specific CPU/cluster on a SoC, the SoC name should not include the CPU

lyakh commented on soc/mediatek/mtk_adsp/soc.c:
```diff
@@ -171,5 +171,9 @@ void c_boot(void)
 	__asm__ volatile("wsr %0, INTCLEAR" :: "r"(val));
 
+	/* Default console, a driver can override this later */
+	void __stdout_hook_install(int (*hook)(int));
```
`#include <zephyr/sys/libc-hooks.h>`?

andyross commented:
OK, I puzzled out the bits needed to add the (IMHO, still needless) cpucluster to the soc tree.  Please see if this matches what you expect.

I will say though, the process... wasn't smooth.  Things like "CONFIG_SOC value does not match string defined in soc.yml" or "Cannot find soc directory" should be obvious error messages and instead are inscrutable secondary build failures.  I spent a lot more time than I'd have expected stuffing `message(WARNING ...)` lines into cmake files.  Also... the DTS filename changes.  Why?  The DTS file is a board asset, not a SOC thing.  It lives in the board directory.  It used to be named the same as the board, but now it isn't, because of something that changed in the SOC layer?

(Also rebased around a collision with llext support having been added to mt8195)

nordicjm commented on boards/mediatek/mt8188_adsp/board.yml:
```diff
@@ -2,4 +2,4 @@ boards:
     vendor: mediatek
     socs:
-      - name: mt8188_adsp
+      - name: mt8188
```
Changes are OK but you could alternatively rename the board to mt8188, then the filenames would be mt8188_adps.dts as it was before, and means your minimal build string would be `west build -b mt8188//adps`. Please squash the changes to the socs/boards added in earlier commits into the earlier commits

andyross replied:
I'd really rather have it be a separate patch if it's OK with you, for symmetry with the older already-merged device that is being changed and just to have a record I know I can look back to.

And the name still creeps me out.  Again, this is **not** an integration for the whole laptop chip.  It's just audio firmware.

nordicjm replied:
How about if you flip the order of converting existing boards to the new format as the first commit then have the new commits adding the final board names in the commits without having a fixup commit at the end?

andyross replied:
Does that really rise to a -1?  It's a non-trivial rebase, and I personally like the order here as it makes the "complicated" change obvious.

nordicjm replied:
You are adding a commit https://github.com/zephyrproject-rtos/zephyr/pull/81154/commits/04ed839c1d989d752364f2259234aba4c85aa0ff and then fixing it in the next commit https://github.com/zephyrproject-rtos/zephyr/pull/81154/commits/eab814b5b8f4ff4b49635c7f41b3d8ac0c1a8f5f so yes. It's fine to have the commits in that order, but the first one should be changed to add the new boards with the correct names outright, then the second commit for updating existing boards, there is no point to adding a commit which has no use, is wrong and is fixed up in a later commit.

andyross replied:
I'm going to try one more time: I really don't want to do that, because if I do that then I'm not going to have a reference to go back to explaining how I did this the first time.  I understand you view that last commit as "obvious" renaming, but I'm not so talented and had to spend a lot of effort to fix things based on my pre-existing understanding of how zephyr board/soc integration works.  I'd like to know I can get back to it to see how it worked.

Basically, can you pretty-please remove your -1 here?  I really don't think commit order should be a merge-blocking kind of thing.  It's a big project and we all need to recognize that different people have different aesthetic senses of how a git log should look.

nordicjm replied:
We have a whole porting guide here: https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html#hw-model-v2
We have an automatic porting script here: https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/utils/board_v1_to_v2.py
We have a whole branch here showing conversions we did for in-tree boards: https://github.com/zephyrproject-rtos/zephyr/tree/collab-hwm
Why do we need a new board to be added that, as part of the same PR, adds the board wrongly in the first commit by not using the things specifically defined for hwmv2 and also should anyone want to bisect a failure for this board would result in them having to use a different board name if they get to the first commit?

If you think the zephyr documentation is lacking, you can open an issue about it or submit a PR improving it, if you want to reference to see how you ported something then you can host that yourself locally on or github or anywhere

andyross replied:
> adds the board wrongly in the first commit by not using the things specifically defined for hwmv2

Yeah, I don't want this to become a big monkey fight, but that's distressingly overstated.  In fact I checked: there's nothing in the existing docs that demands the use of a cpucluster, it just talks about what the syntax means.  The demand for particular structure was **you're personal design aesthetic**.  And... that's fine!  Your code, your model, you get to make demands like that.

But it's not correct to characterize the early patches as "incorrect" either, given that they matched the idioms used by many (probably most) existing boards in the tree.  So if you do make those demands, part of the job of being a good maintainer is staying with the PR as developers push back with questions and attempts at compromise and not walking away in pique.

nordicjm replied:
If there are other boards not complying then they should be fixed, these are not boards/socs I use so I don't know of any

nordicjm commented on soc/mediatek/mtk_adsp/mtk_adsp_load.py:
```diff
@@ -102,40 +193,35 @@ def main():
             maps["sram"][i] = sram[i]
-        for i in range(sram_len, MAPPINGS["sram"][1]):
-            maps["sram"][i] = 0
+        #for i in range(sram_len, mmio["sram"][1]):
+        #    maps["sram"][i] = 0
```
is this not needed?

andyross replied:
Zephyr will zero its own bss after boot, so no.  It's a hygiene thing only (to guarantee that all memory looks consistent, even things like .noinit sections), but disabled because on one of the boards I have it appears to glitch the bus and reboot the whole device.

nashif commented on soc/mediatek/mtk_adsp/soc.yml:
```diff
@@ -1,6 +1,21 @@
 family:
   - name: mtk_adsp
```
with all those changes, is the family still correct? mt8195 is the container of adsp 'cpucluster', so the family can't be named mtk_adsp, no?

andyross replied:
Indeed, arguably no, for the same reason as the board quibble above.  But again, this isn't[1] intended to be a framework for all Zephyr firmware to run on any MediaTek device.  It's just supposed to be audio firmware!

But OK.  It doesn't really hurt anything to change names.  Will respin again.

[1] And really can't be.  Not only am I not working on that stuff, it bears mentioning that I don't actually work for MediaTek and am only "authorized"[2] to be doing this in the context of "port largely-pre-existing SOF audio firmware to Zephyr" and not "Design a Zephyr integration for all MediaTek devices". 

[2] There's no formal authorization involved, the SOF code is mostly upstream (though I'm actually ahead of that process on 8196) and Apache-licensed.  But I absolutely don't want to stomp on any toes.

andyross committed:
soc/mediatek: Add missing snippets sections

This got missed by the first version, but recent SOF needs it for
STRUCT_SECTION_ITERABLE usage.

Signed-off-by: Andy Ross <andyross@google.com>

andyross commented:
Rebase around another collision.  Should be good to go now, please review.  Also note I snuck in one final 1-line patch to the linker script.

dcpleung commented on arch/Kconfig:
```diff
@@ -579,11 +575,12 @@ config IRQ_OFFLOAD
 config IRQ_OFFLOAD_NESTED
 	bool "irq_offload() supports nested IRQs"
 	depends on IRQ_OFFLOAD
+	default y if ARM64 || X86 || RISCV || XTENSA
```
The side effect of this is that you can turn it off at build time, e.g. `west build`, which I think is not quite the correct behavior. Maybe we can leave the other arch to select this kconfig. But for Xtensa, this is done in the SoC layer?

andyross replied:
I think that's sort of the point.  Xtensa is the first arch to hit this edge case.  What's happening here is that we have an architecture which can support nested interrupts, but an instantiation that doesn't actually have a high priority software interrupt set up that irq_offload() can use, so in practice we can't do it.  That's likely to be a situation lots of devices could hit in practice.

Basically it was too hard-coded, and needs to be a little softer IMHO.  And in any case irq_offload() really only exists for test purposes anyway.

andyross commented:
Another ping for review.  Again, should be good to merge unless I've missed something.

iuliana-prodan reviewed (dismissed):
LGTM

andyross commented:
@nordicjm you still have a -1 on this.  Are we good now?  If not, come back and engage in combat?

nashif commented:
giving owners of dismissed reviews some time before approving that myself and getting this merged.

nashif commented:
![image](https://github.com/user-attachments/assets/d8da60c5-629a-4f70-a171-0191f4d22e1a)

Why? Please provide some details here, I am not following why this is a meeting/arch review topic.

nordicjm commented:
> ![image](https://private-user-images.githubusercontent.com/180017/389903433-d8da60c5-629a-4f70-a171-0191f4d22e1a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI2MTcxNjEsIm5iZiI6MTczMjYxNjg2MSwicGF0aCI6Ii8xODAwMTcvMzg5OTAzNDMzLWQ4ZGE2MGM1LTYyOWEtNGY3MC1hMTcxLTAxOTFmNGQyMmUxYS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyNlQxMDI3NDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01YzFiZTc2YzE3N2ExNjE2ODhjMTU0YTQ5NjUwZjRjOTY3NmE0MDk1ODU5NWFiN2U4YTdkNGM3NWJhZjc3YjIyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.QfFGkJwegwZm_fKIp19NHDO4mMSERsU7nQ6x_fOV9xc)
> 
> Why? Please provide some details here, I am not following why this is a meeting/arch review topic.

You dismissed the nack, I've provided the reasons as to why I don't agree here: https://github.com/zephyrproject-rtos/zephyr/pull/81154#discussion_r1858038922 and am following the following the escalation process as you said

nashif commented:
> You dismissed the nack, I've provided the reasons as to why I don't agree here: [#81154 (comment)](https://github.com/zephyrproject-rtos/zephyr/pull/81154#discussion_r1858038922) and am following the following the escalation process as you said

ok, missed that, GH just puts it in the middle somewhere.

nashif reviewed (changes requested):
@andyross  you are adding a few "boards" but I do not see them come with a yaml file for twister and testing.

Also, I fail to build one of the boards with plain west.


```west twister  -p mt8195/mt8195/adsp -T samples/hello_world```

```/home/nashif/zephyrproject/modules/hal/xtensa/include/xtensa/config/core.h:51:10: fatal error: xtensa/config/core-matmap.h: No such file or directory
   51 | #include <xtensa/config/core-matmap.h>```

andyross commented:
That's right, these guys still require the Cadence toolchain.  Though 8195 has one in the SDK already, that one should work.  There are instructions for building a gcc in the board docs.  IIRC the last time this got discussed there was a general decision that there were too many Xtensa toolchains in the SDK and that we should just be supporting a subset.  But I can submit the two needed if that's required.

And @nordicjm: you're really going to the mattresses over commit order?  I mean, I guess we can do that.  But can you clarify that you're OK with the structure of the board integration here, you just want to see the commit order changed?

nordicjm commented on boards/mediatek/mt8186/mt8186_adsp.dts:
```diff
@@ -0,0 +1,82 @@
+
+/dts-v1/;
+/ {
+
```
missing indent below here

andyross replied:
This comment got made earlier and the response was that it was intentional, for the same reason you don't indent inside a `namespace { }` directive in C++.   I thought you were OK with that?

nordicjm replied:
I've only ever seen it be indented on boards e.g. https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/nordic/nrf52840dk/nrf52840dk_nrf52840.dts I can't say I've seen a board where it's not indented, and for the C++ code bases I have the namespace {} is indented

andyross replied:
So you're not OK with that then?  Will fix.

nordicjm commented on boards/mediatek/mt8188/mt8188_adsp.dts:
```diff
@@ -0,0 +1,82 @@
+/dts-v1/;
+/ {
+
+#address-cells = <1>;
```
as above

nordicjm commented on boards/mediatek/mt8196/mt8196_adsp.dts:
```diff
@@ -0,0 +1,109 @@
+/dts-v1/;
+/ {
+
+#address-cells = <1>;
```
as above

nordicjm commented on soc/mediatek/mt8xxx/Kconfig.defconfig:
```diff
@@ -23,11 +23,22 @@ config MAX_IRQ_PER_AGGREGATOR
 config XTENSA_TIMER
 	default n
 
+config SYS_CLOCK_HW_CYCLES_PER_SEC
+	default 13000000
```
awaiting comment from Gerard

gmarull replied:
see e.g.

https://github.com/zephyrproject-rtos/zephyr/blob/b7b4de8afeb93b480b6a8e3bef18c76ea6a82bc9/soc/gd/gd32/gd32f4xx/Kconfig.defconfig.gd32f470#L6-L7

andyross replied:
Alas, this doesn't seem to work.  This is what I have right now, and it evaluates to a zero instead of 13M.   Variants like specifying the path explicitly (which is a non-starter as the node contains a reg address which we obviously don't want to be copying into kconfig), using an dt_node_array_prop_int with an index of zero, etc... Any guidance?  I actually don't see any other drivers in the tree pulling an integer out of a nodelabel like this.  The `cpus@0` idiom is common, but this isn't a CPU frequency, it's owned by the timer device.

```diff
diff --git a/boards/mediatek/mt8196/mt8196_adsp.dts b/boards/mediatek/mt8196/mt8196_adsp.dts
index 1cc43225f70..e4917d72152 100644
--- a/boards/mediatek/mt8196/mt8196_adsp.dts
+++ b/boards/mediatek/mt8196/mt8196_adsp.dts
@@ -79,6 +79,7 @@ soc {
        ostimer64: ostimer64@1a00b080 {
                compatible = "mediatek,ostimer64";
                reg = <0x1a00b080 28>;
+               clock-frequency = <13000000>;
        };
 
        ostimer0: ostimer@1a00b000 {
diff --git a/soc/mediatek/mt8xxx/Kconfig.defconfig b/soc/mediatek/mt8xxx/Kconfig.defconfig
index b2796b8889d..1f28b2317e5 100644
--- a/soc/mediatek/mt8xxx/Kconfig.defconfig
+++ b/soc/mediatek/mt8xxx/Kconfig.defconfig
@@ -37,7 +37,7 @@ config XTENSA_TIMER
        default n
 
 config SYS_CLOCK_HW_CYCLES_PER_SEC
-       default 13000000
+       default $(dt_node_int_prop_int,$(dt_nodelabel_path,ostimer64),clock-frequency)
 
 config MAIN_STACK_SIZE
        default 2048
```

nordicjm commented:
> And @nordicjm: you're really going to the mattresses over commit order? I mean, I guess we can do that. But can you clarify that you're OK with the structure of the board integration here, you just want to see the commit order changed?

I'm not bothered about commit order, just that the commit adding the new board should add the new board with the correct name in that commit, not with the old name and having it changed in another commit (the other commit changing existing board/soc names is fine to be separate). Have had a quick look over `files changed` instead of individual commits and can see nothing major just some minor nits, awaiting comment from Gerard on one point if it should come from dts of if Kconfig is fine.

andyross commented:
OK, well can we maybe hit the gas on new review items?  This has been up for three weeks now and I've got work backing up behind it.  Will address the SDK/HAL patches needed for gcc builds and the twister thing this morning as soon as I can find time.

nashif commented:
> That's right, these guys still require the Cadence toolchain.

that is something you can put in the yaml file, list the toolchains supported: xcc, xt-clang or zephyr if zephyr sdk is supported.

andyross commented:
SDK and HAL patches up at https://github.com/zephyrproject-rtos/sdk-ng/pull/842 and https://github.com/zephyrproject-rtos/hal_xtensa/pull/34 (though I'm mixed on whether to add more toolchains to the SDK, seems like it's just as simple to use the SOF instructions to build your own out of thesofproject/xtensa-overlays for the rare users who want open source tooling), will update the code here in a few hours, have a shopping trip booked for the afternoon.

andyross committed:
boards/mediated: Add twister.yaml

Add a twister YAML file for these devices.  Right now they're flagged
as xt-clang only, though xcc and SOF-derived gcc toolchains are known
to work too.  We'll enable the zephyr SDK once sdk-ng support lands.

Signed-off-by: Andy Ross <andyross@google.com>

andyross commented:
Giant rebase finished, the big file and kconfig rename is now upstream of the mt8196 patch, and folded into the addition of 818x.  Twister added.  New comments addressed.  Can I ask folks to please get back promptly with more criticism and replies?  I really do need this merged, and am a little shocked at how difficult the process has become.

andyross commented:
Also can someone provide guidance on the sudden appearance of these "ruff" failure in the python scripts?  This code is almost all in-tree already, it's just being touched by the PR to add a few lines.  And some of the failures seem... questionable.  It's NAKing a script because the imports statements... aren't sorted?!

andyross commented:
> It's NAKing a script because the imports statements... aren't sorted?!

Even better, apparently we can now no longer submit a change to a python script if it happens to include a hex constant with (gasp) lower case digits.  I'll fix the scripts in this PR out of necessity and malicious compliance, but something seriously needs to change with this CI rule; it's not going to scale at all.

andyross committed:
soc/mediatek: Ruffify python scripts

These in-tree scripts fail the new ruff checks.  Clean things up so
modifications can merge.

Signed-off-by: Andy Ross <andyross@google.com>

kartben commented:
> > It's NAKing a script because the imports statements... aren't sorted?!
> 
> Not even "not sorted alphabetically". It has its own ideas about package order that it doesn't seem to document.
> 
> Even better, apparently we can now no longer submit a change to a python script if it happens to include a hex constant with (gasp) lower case digits. I'll fix the scripts in this PR out of necessity and malicious compliance, but something seriously needs to change with this CI rule; it's not going to scale at all.

@andyross if there are scripts that you're including verbatim from other projects, like it looks to be the case based on copyright notices, it would be fine to just add them to ruff-excludes.toml IMO. cc @pdgendt

andyross committed:
soc/mediatek/adsp: Source timer rate from DTS

These devices have an architecturally fixed 13 MHz clock device.  But
thankfully you can put a default into a DTS binding so we don't have
to repeat it for all of them.

Signed-off-by: Andy Ross <andyross@google.com>

andyross commented:
No, they're my scripts.  Just not new in this PR yet still failing on ruff checks because like most linters it doesn't understand the idea of "only check the deltas" (which is a really hard problem).  It's the standard complaint about formatters: if the existing source isn't valid already, they have the effect of disallowing all changes.

andyross commented:
LOL.  I've iteratively fixed the ruffination errors I can duplicate.  Now it's failing with:

```
 Python format error:Run 'ruff format soc/mediatek/mt8xxx/gen_img.py'
File:soc/mediatek/mt8xxx/gen_img.py
 Python format error:Run 'ruff format soc/mediatek/mt8xxx/mtk_adsp_load.py'
File:soc/mediatek/mt8xxx/mtk_adsp_load.py
Compliance error, check for error messages in the "Run Compliance Tests" step
You can run this step locally with the ./scripts/ci/check_compliance.py script.
```

... there's no error there!  What do I fix?!

kartben commented:
> LOL.  I've iteratively fixed the ruffination errors I can duplicate.  Now it's failing with:
> 
> ```
>  Python format error:Run 'ruff format soc/mediatek/mt8xxx/gen_img.py'
> File:soc/mediatek/mt8xxx/gen_img.py
>  Python format error:Run 'ruff format soc/mediatek/mt8xxx/mtk_adsp_load.py'
> File:soc/mediatek/mt8xxx/mtk_adsp_load.py
> Compliance error, check for error messages in the "Run Compliance Tests" step
> You can run this step locally with the ./scripts/ci/check_compliance.py script.
> ```
> 
> ... there's no error there!  What do I fix?!

Didn't realize you moved the files. So yeah, the ruff-excludes.toml entries needed to be updated after all :) which would have had the original intended behavior of not complaining about existing files.
Not sure about the remaining error though :/ I guess I would try to run the ruff format commands locally to see if they output any error that the CI script would not be showing

andyross commented:
Ooph.  Got it to work.  What that text was trying to tell me is that "ruff format" reported that it had made changes to the file, but none that it logged.  You can't just run ruff and fix what it complains about, you need to run ruff and **commit** what it **doesn't** complain about. :)

andyross commented:
Sigh, unbelievable.  Yes, indeed I did move the files as part of the big record demanded above for hwmv2 compliance.  But obviously my brain didn't file that as a "move" since it was just a minor change request trying to get this thing merged.  And in any case I guess I didn't realize there was a pre-existing exclude list added for scripts already in-tree (FWIW: git understands the idea of file motion, it wouldn't be impossible to write a checker that could update the .toml based on the PR being examined).

This... really isn't supposed to be this difficult.  There was a time when you could add a new board to Zephyr with half an hour of work and someone would merge it instantly.  Because it's a **new board**;  it's by definition not going to break anything and you can always fix it later iteratively.  Now we're a giant tangle of fiefdoms and processes and only-mostly-documented idioms, and the only tool anyone has to exert control is the -1 button in GitHub.

andyross commented:
Fixed indentation in DTS.  Any other complaints?  Would be really, really nice to see a few +1's and get this merged before people start to disappear around the holidays.

Just as an aside @nordicjm, as far as the C++ namespace point, you're definitely in the minority there.  Consensus is generally that "needless enforced top-level indentation" is a bad thing and most of the community has disallowed it.  See these two popular style guides, for example:

https://llvm.org/docs/CodingStandards.html#namespace-indentation
https://google.github.io/styleguide/cppguide.html#Namespace_Formatting

Also clang-format won't do it and AFAICT can't be made to; you can see the examples in their docs: https://clang.llvm.org/docs/ClangFormatStyleOptions.html

I don't think it's really that much a stretch to imagine something similar in DTS context, the reason is identical (there's a top-level scope that is enforced by syntax and can't be avoided, so effectively the whole file gets indented for no particularly good reason).  Something to consider for the future when we've all calmed down a bit.

nordicjm reviewed (approved):
Thank you for updating

nordicjm commented:
> Just as an aside @nordicjm, as far as the C++ namespace point, you're definitely in the minority there. Consensus is generally that "needless enforced top-level indentation" is a bad thing and most of the community has disallowed it. See these two popular style guides, for example:

Interesting, was not aware of those. The codebase is so old now that I don't remember if it was the IDE that added the indentation or if I did, presumably it was me...
```
namespace Ui
{
    class AutMainWindow;
}
```
Just created a new project in the IDE and must have been me that indented it
```
namespace Ui {
class MainWindow;
}
```

kartben reviewed (commented):
I won't make this a blocker but the following will need to be addressed as follow-ups please:

- documentation as per the comments ; 
- `mt8195_adsp` board being renamed to `mt8195` will need to be mentioned in the 4.1 migration guide

andyross commented on boards/mediatek/index.rst:
```diff
@@ -0,0 +1,166 @@
+.. _boards-mtk_adsp:
+
+Mediatek Audio DSPs
+###################
+
```
Heh.  Well I do need to point out that when that patch was written, that directory **was** exclusively to Audio DSPs.   :)  But sure, will figure it out.  Presumably I can have stub pages and link to the shared one or whatever.

ajf58 commented on boards/mediatek/index.rst:
```diff
@@ -0,0 +1,166 @@
+
+.. code-block:: console
+
+   user@dev_host:~$ west build -b mt8186//adsp samples/hello_world
```
```suggestion
   user@dev_host:~$ west build -b mt8186/adsp samples/hello_world
```
nit/not a blocker, but is there an extra forward slash that crept in?

ajf58 commented on boards/mediatek/index.rst:
```diff
@@ -0,0 +1,166 @@
+**********
+
+The MT8195 toolchain is already part of the Zephyr SDK, so builds for
+the ``mt8195//adsp`` board should work out of the box simply following
```
```suggestion
the ``mt8195/adsp`` board should work out of the box simply following
```
nit: As  above, not a blocker, but has a double slash crept in?

andyross replied:
Nope, see above. :)  That's the real name, it's shorthand for "mt8195/mt8195/adsp"

ajf58 commented on boards/mediatek/mt8188/mt8188_adsp.dts:
```diff
@@ -0,0 +1,81 @@
+			interrupt-parent = <&core_intc>;
+		};
+
+		ostimer64: ostimer64@10b83080 {
+			compatible = "mediatek,ostimer64";
+			reg = <0x10b83080 28>;
+		};
+
+		ostimer0: ostimer@10b83000 {
+			compatible = "mediatek,ostimer";
+			reg = <0x10b83000 16>;
+			interrupt-parent = <&core_intc>;
+			interrupts = <18 0 0>;
+		};
```
nit: could/should these entries be in reversed order, given their addresses?

andyross replied:
They're not the same kind of device.  The first is the 13MHz 64 bit upcounter used for system timing, the second is a 32 bit 26MHz down-counting interrupt source.  The timer driver uses them both, but in fact the second has multiple instantiations on the chips (though not used by Zephyr), so they make sense to declare separately.  As far as order: my sense of aesthetics likes having the master counter first, I guess.

ajf58 commented on boards/mediatek/mt8195/mt8195_adsp.dts:
```diff
@@ -0,0 +1,94 @@
+			interrupt-parent = <&core_intc>;
+		};
+
+		ostimer64: ostimer64@1080d080 {
+			compatible = "mediatek,ostimer64";
+			reg = <0x1080d080 28>;
+		};
+
+		ostimer0: ostimer@1080d000 {
+			compatible = "mediatek,ostimer";
+			reg = <0x1080d000 16>;
+			interrupt-parent = <&intc23>;
+			interrupts = <11 0 0>;
+		};
```
nit: As above, should these nodes be the other way round?

ajf58 commented on boards/mediatek/mt8196/mt8196_adsp.dts:
```diff
@@ -0,0 +1,109 @@
+			interrupt-parent = <&core_intc>;
+		};
+
+		ostimer64: ostimer64@1a00b080 {
+			compatible = "mediatek,ostimer64";
+			reg = <0x1a00b080 28>;
+		};
+
+		ostimer0: ostimer@1a00b000 {
+			compatible = "mediatek,ostimer";
+			reg = <0x1a00b000 16>;
+			interrupt-parent = <&intc_g1>;
+			interrupts = <8 0 0>;
+		};
```
nit: As above.

ajf58 reviewed (approved):
LGTM.

I've raised a couple of nits, but nothing that should gate this being merged onto `main`.
